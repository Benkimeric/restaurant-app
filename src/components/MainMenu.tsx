import React, { useState } from "react";

import menu from "../mock/menu.json";

const MainMenu = () => {
  const [state, setState] = useState<any>(menu);

  const handleClick = (name: String) => {
    setState((state: any) =>
      state.map((item: any) => {
        let val = item;
        if (item.name === name) {
          val = {
            ...val,
            expand: val.expand ? false : true,
            expandChoices: val.expand ? false : val.expandChoices,
            related: val.related.map((related: any) => ({
              ...related,
              expandRelated: val.expand ? false : related.expandRelated
            }))
          };
        }

        return val;
      })
    );
  };

  const otherChoicesUnchecked = (name: any, choices: any): boolean => {
    const checkedChoices = choices.filter((choice: any) => {
      if (choice.checked && choice.name !== name) {
        return true;
      }
      return false;
    });

    return checkedChoices.length === 0;
  };

  const clickChoice = (name: string, choiceName: string) => {
    setState((state: any) =>
      state.map((item: any) => {
        let val = item;
        if (item.name === name) {
          const othersUnchecked = otherChoicesUnchecked(
            choiceName,
            val.choices
          );

          val = {
            ...val,
            expandChoices: val.expandChoices && othersUnchecked ? false : true,
            related: val.related.map((related: any) => ({
              ...related,
              expandRelated:
                val.expandChoices && othersUnchecked ? false : val.expandRelated
            })),
            choices: val.choices.map((choice: any) => {
              if (choice.name === choiceName) {
                return {
                  ...choice,
                  checked:
                    choice.checked && choice.name === choiceName
                      ? false
                      : choice.name === choiceName
                };
              }

              return choice;
            })
          };
        }

        return val;
      })
    );
  };

  const openRelated = (choiceName: string, relatedName: string) => {
    setState((state: any) =>
      state.map((item: any) => {
        let value = item;
        let related = [];

        if (value.name === choiceName) {
          related = value.related.map((related: any) => {
            if (related.name === relatedName) {
              return {
                ...related,
                expandRelated: related.expandRelated ? false : true
              };
            }
            return related;
          });
        }

        return { ...value, related };
      })
    );
  };

  return (
    <div>
      <h1>Restaurant Menu</h1>
      {state.map((menu: any, idx: any) => {
        return (
          <>
            <div key={idx}>
              <label htmlFor={menu.name}>{menu.name}</label>
              <input
                id={"mainChecked"}
                type="checkbox"
                onClick={() => handleClick(menu.name)}
              />
            </div>

            {/* choices */}
            {menu.expand &&
              menu.choices.map((choice: any) => {
                return (
                  <div className="choices">
                    <label>{choice.name}</label>
                    <input
                      id="showRelated"
                      type="checkbox"
                      onClick={e => clickChoice(menu.name, choice.name)}
                    />
                  </div>
                );
              })}

            {/* related names*/}
            {menu.expandChoices && menu.related.length > 0 && (
              <p className="related-choices">You may also want:</p>
            )}
            {menu.expandChoices &&
              menu.related.map((choice: any) => {
                return (
                  <div className="related-choices">
                    <label className="choices">{choice.name}</label>
                    <input
                      id={choice.name}
                      type="checkbox"
                      onClick={() => openRelated(menu.name, choice.name)}
                    />
                    {/* related choices */}
                    {choice.expandRelated &&
                      choice.choices.map((relatedChoice: any) => {
                        return (
                          <div className="related-choice">
                            <label className="">{relatedChoice.name}</label>
                            <input id={relatedChoice.name} type="checkbox" />
                          </div>
                        );
                      })}
                  </div>
                );
              })}
          </>
        );
      })}
    </div>
  );
};

export default MainMenu;
