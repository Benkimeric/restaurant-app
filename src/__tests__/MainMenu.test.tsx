import React from "react";
import Enzyme, { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import MainMenu from "../components/MainMenu";

Enzyme.configure({ adapter: new Adapter() });

describe("Tests for main menu component", () => {
  const wrapper = mount(<MainMenu />);

  it("should render main menu component without fail", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should display the title on load", () => {
    const title = wrapper.find("h1");
    expect(title.contains("Restaurant Menu")).toEqual(true);
  });

  it("should display main menu", () => {
    expect(wrapper.contains("Salad")).toEqual(true);
    expect(wrapper.contains("Entree")).toEqual(true);
    expect(wrapper.contains("Soup")).toEqual(true);
  });

  it("on click it should expand menu", () => {
    const checkbox = wrapper.find("#mainChecked").first();
    checkbox.simulate("click");
    expect(wrapper.contains("Greek")).toEqual(true);
  });

  it("it should open related menu name on click", () => {
    const checkbox = wrapper.find("#showRelated").first();
    checkbox.simulate("click");
    expect(wrapper.contains("Dressing")).toEqual(true);
  });

  it("it should open related menu choices on menu name click", () => {
    const checkbox = wrapper.find("#Dressing").first();
    checkbox.simulate("click");
    expect(wrapper.contains("Italian")).toEqual(true);
  });
});
