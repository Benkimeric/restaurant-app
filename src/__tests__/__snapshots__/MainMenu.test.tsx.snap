// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for main menu component should render main menu component without fail 1`] = `ReactWrapper {}`;
